### ISSUE TITLE
-compdb and args with spaces

### ISSUE THREAD

**Issue by g3cko** [fastbuild/fastbuild#741]
Hi there,

I think this is just a bug but I was having some issues using "-compdb" (which has been super helfpul!) and a default install of VS (path to "Program Files" with spaces). 

What was happening:   the resulting args in the json file are all jumbled up.  

Results in json:
`"/imsvcC:/Program", "Files", "(x86)/Microsoft", "Visual", "Studio/2019/Professional/VC/Tools/MSVC/14.25.28610/include/ .... <further args continued>`   

Expected and observed with code below:
`/imsvc\"C:/Program Files (x86)/Microsoft Visual Studio/2019/Professional/VC/Tools/MSVC/14.25.28610/include/\"`

It appears that this line is not needed?  

```
void CompilationDatabase::HandleObjectListNode( ObjectListNode * node )
...
 //Unquote( argument );   <-----------
 JSONEscape( argument );
```

**Comment by ffulin** [comment]
I think the intent of the quote removal is to correctly handle cases such as:
```C++
Arg1 "-IPath With Spaces" Arg3
```
This should be tokenized as 3 arguments:
* Arg1
* -IPath With Spaces
* Arg3

Each argument is then written to the json surrounded in quotes. Without removal of the quotes from the second argument, it would also contain quotes. Although the inner quotes would be escaped to make valid json, the quotes are not part of the argument (would not be passed to the external process) and as such should not be present in the json.

With your example, and looking at the tokenization code, I think the problem probably arises from args the contain quotes internally like this:
```C++
Arg1 -I"Path With Spaces" Arg3
```
I think because the second arg doesn't begin with a space, it's tokenized incorrectly as 5 arguments:
* Arg1
* -I"Path
* With
* Spaces"
* Arg3

It's likely a smarter tokenization function is required to correctly handle both cases.

In the interim, you could probably work around the problem by moving the quotes around the entire arg:
```C++
"/imsvcC:/Program Files (x86)/Microsoft Visual Studio/2019/Professional/VC/Tools/MSVC/14.25.28610/include/"
```
or perhaps as two args (I'm not sure what clang expects/accepts):
```C++
"/imsvc" "C:/Program Files (x86)/Microsoft Visual Studio/2019/Professional/VC/Tools/MSVC/14.25.28610/include/"
```

**Comment by g3cko** [comment]
>Without removal of the quotes from the second argument, it would also contain quotes.  [...] the quotes are not part of the argument

I guess that is what I don't understand, when do you want this behavior?  The quotes are part of the arg, otherwise this is not going to compile 

`cl.exe ...  -IPath With Spaces ... `  

I did find that I can double quote using one of each single and double (otherwise Unquote will strip all of the same type) to work around this without changing the src so that is good so I can keep upgrading via the build downloads.

**Comment by ffulin** [comment]
>The quotes are part of the arg, otherwise this is not going to compile

That's the thing: the quote are actually not part of the arg. They are special tokens used by the invoker to handle spaces in the args.

Consider this application:
```C++
int main(int argc, char** argv)
{
    return 1;
}
```

If you compile and run that, from the command line:
```
test.exe -I"Arg With Spaces"
```

At runtime, your application sees the following single argument:
```
-IArg With Spaces
```
The quotes are nowhere to be seen by the time they get to your application.

Depending on the what API the OS provides to invoke an executable, you either need to provide these quotes or not.  (On windows, you provide a single command line string and must quote, but on OSX/Linux you provide an array of args and must not quote (otherwise you are adding quotes into the args that should not be there)).

With that in mind, it makes sense to me the at compdb json contains the actual args, free of this "escaping", and leaves it to the caller to do the right thing based on context.

**Comment by ffulin** [comment]
To clarify: There is a bug here, but it's not that the args are being removed. It's that they are *not* being removed when the quote is not the first character. i.e. FASTBuild is not handling args consistently with the cmd.exe or other applications.

**Comment by ffulin** [comment]
There is an existing related issue: https://github.com/fastbuild/fastbuild/issues/510
Specifically, parts of FASTBuild elsewhere make similar bad assumptions about how quotes should be handled as the compdb code we're discussing.

**Comment by g3cko** [comment]
Ahh - I see what you mean now.  I wasn't making a clear distinction between what main sees, and what I see in a cmd.exe prompt invoking that app.  

I primary dev from Windows, didn't realize on OSX/Linux you provide an array of args.  Thanks for clarifying that.

**Comment by ffulin** [comment]
Various arg parsing fixes have been implemented in https://github.com/fastbuild/fastbuild/commit/36a8b4dbce16c930afcd14210c80a69529ec3ada

This does not directly fix the behavior of -compdb as reported in this issue, but is the first part of a fix as now:
a) args are handled consistently on all platforms
b) the necessary functionality now exists in code to allow the -compdb output to be generated correctly

**Comment by ffulin** [comment]
In addition to the fixes to general arg parsing mentioned above, some additional fixes to arg handling for -compb specifically have now been applied in https://github.com/fastbuild/fastbuild/commit/fe10e5a12bb8ee282425dbb00c447f0a099cce24.

These fixes combined resolve the issue in this thread.

The fixes will be in the next release (v1.13). I'll keep this issue open until that version is released.

**Comment by ffulin** [comment]
v1.13 has now been released and includes this fix. Thanks again for reporting the bug.
